geopandas/geodataframe.py

```diff
@@ -1624,6 +1624,15 @@ def apply(self, func, axis=0, raw=False, result_type=None, args=(), **kwargs):
     def _constructor(self):
         return _geodataframe_constructor_with_fallback
 
+    def _constructor_from_mgr(self, mgr, axes):
+        # replicate _geodataframe_constructor_with_fallback behaviour,
+        # unless safe to skip
+        if not any(isinstance(block.dtype, GeometryDtype) for block in mgr.blocks):
+            return _geodataframe_constructor_with_fallback(
+                pd.DataFrame._from_mgr(mgr, axes)
+            )
+        return GeoDataFrame._from_mgr(mgr, axes)
```

Comment:
Note we can't just have 

`return _geodataframe_constructor_with_fallback(pd.DataFrame._from_mgr(mgr, axes))`
always, test_sindex.py::test_rebuild_on_row_slice fails because copys of GeometryArrays don't have a way to retain the sindex